core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java

```diff
@@ -0,0 +1,216 @@
+ * provided services file and leave the standard factories and under file system implementations
+ * unavailable.
+ * </p>
+ * <h4>Manual Registration</h4>
+ * <p>
+ * To manually register a factory simply pass an instance of your factory to the
+ * {@link #add(UnderFileSystemFactory)} method. This can be useful when your factory cannot be
+ * instantiated without arguments or in cases where automatic discovery does not give your factory
+ * priority. Factories registered this way will be registered at the start of the factories list so
+ * will have the first opportunity to indicate whether they support a requested path.
+ * </p>
+ */
+public class UnderFileSystemRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final List<UnderFileSystemFactory> FACTORIES =
+      new CopyOnWriteArrayList<UnderFileSystemFactory>();
+  private static boolean sInit = false;
+
+  static {
+    // Call the actual initializer which is a synchronized method for thread safety purposes
+    init();
+  }
+
+  static synchronized void init() {
+    if (sInit) {
+      return;
+    }
+
+    // Discover and register the available factories
+    ServiceLoader<UnderFileSystemFactory> discoverableFactories =
+        ServiceLoader.load(UnderFileSystemFactory.class);
+    Iterator<UnderFileSystemFactory> iter = discoverableFactories.iterator();
+    while (iter.hasNext()) {
+      UnderFileSystemFactory factory = iter.next();
+      LOG.debug("Discovered Under File System Factory implementation {} - {}", factory.getClass(),
+          factory.toString());
+      FACTORIES.add(factory);
+    }
+
+    sInit = true;
+  }
+
+  /**
+   * Resets the registry to its default state
+   * <p>
+   * This clears the registry as it stands and rediscovers the available factories.
+   * </p>
+   */
+  public static synchronized void reset() {
+    // Can't reset if we're not initialized
+    if (!sInit) {
+      return;
+    }
+
+    // Reset state
+    sInit = false;
+    FACTORIES.clear();
+
+    // Reinitialise
+    init();
+  }
+
+  /**
+   * Adds a new factory
+   * <p>
+   * Factories are added to the start of the factories list so they can override the existing
+   * automatically discovered factories. Generally if you use the ServiceLoader mechanism properly
+   * it should be unnecessary to call this, however since ServiceLoader discovery order may be
+   * susceptible to class loader behavioural differences there may be rare cases when you need to
+   * manually register the desired factory.
+   * <p>
+   * 
+   * @param factory
+   *          Factory to add
+   */
+  public static void add(UnderFileSystemFactory factory) {
+    if (factory == null) {
+      return;
+    }
+
+    LOG.debug("Registered Under File System Factory implementation {} - {}", factory.getClass(),
+        factory.toString());
+    FACTORIES.add(0, factory);
```

Comment:
Thanks, can you leave a comment in the code then?
On Feb 6, 2015 3:58 PM, "Rob Vesse" <notifications@github.com> wrote:

> In core/src/main/java/tachyon/underfs/UnderFileSystemRegistry.java
> <https://github.com/amplab/tachyon/pull/647#discussion_r24281776>:
>
> > +   * it should be unnecessary to call this, however since ServiceLoader discovery order may be
> > +   * susceptible to class loader behavioural differences there may be rare cases when you need to
> > +   * manually register the desired factory.
> > +   * <p>
> > +   *
> > +   * @param factory
> > +   *          Factory to add
> > +   */
> > +  public static void add(UnderFileSystemFactory factory) {
> > +    if (factory == null) {
> > +      return;
> > +    }
> > +
> > +    LOG.debug("Registered Under File System Factory implementation {} - {}", factory.getClass(),
> > +        factory.toString());
> > +    FACTORIES.add(0, factory);
>
> The intention of using CopyOnWriteArrayList iwas to avoid having to
> explicitly lock on FACTORIES and not to worry if someone is changing its
> definition while other methods are being called
>
> Implementation wise I intentionally insert at the head because the point
> of this method is to allow users to manually override the automatically
> discovered implementations where necessary.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/amplab/tachyon/pull/647/files#r24281776>.
>